arangod/RocksDBEngine/RocksDBVectorIndexList.cpp

```diff
@@ -55,63 +59,88 @@ RocksDBInvertedListsIterator::RocksDBInvertedListsIterator(
+  }
 }
 
-RocksDBInvertedListsFilteringIterator::RocksDBInvertedListsFilteringIterator(
-    RocksDBVectorIndex* index, LogicalCollection* collection,
-    SearchParametersContext& searchParametersContext, std::size_t listNumber,
-    std::size_t codeSize)
-    : InvertedListsIterator(),
-      _index(index),
-      _collection(collection),
-      _searchParametersContext(searchParametersContext),
-      _listNumber(listNumber) {
-  TRI_ASSERT(searchParametersContext.filterExpression != nullptr);
-  RocksDBTransactionMethods* mthds = RocksDBTransactionState::toMethods(
-      searchParametersContext.trx, collection->id());
+// Explicit instantiations
+template struct RocksDBInvertedListsIterator<NoStoredValuesStrategy>;
+template struct RocksDBInvertedListsIterator<WithStoredValuesStrategy>;
 
-  _rocksdbKey.constructVectorIndexValue(_index->objectId(), _listNumber);
-  _batchIt = mthds->NewIterator(index->columnFamily(), [&](auto& opts) {
-    TRI_ASSERT(opts.prefix_same_as_start);
-  });
-  _batchIt->Seek(_rocksdbKey.string());
+/// RocksDBInvertedListsFilteringIterator
+template<VectorIndexStoredValuesStrategy Strategy>
+RocksDBInvertedListsFilteringIterator<Strategy>::
+    RocksDBInvertedListsFilteringIterator(
+        RocksDBVectorIndex* index, LogicalCollection* collection,
+        SearchParametersContext& searchParametersContext,
+        std::size_t listNumber, std::size_t codeSize)
+    : RocksDBInvertedListsIteratorBase(
+          index, collection, searchParametersContext.trx, listNumber, codeSize),
+      _searchParametersContext(searchParametersContext) {
+  TRI_ASSERT(searchParametersContext.filterExpression != nullptr);
   skipOverFilteredDocuments();
 }
 
-[[nodiscard]] bool RocksDBInvertedListsFilteringIterator::is_available() const {
+template<VectorIndexStoredValuesStrategy Strategy>
+[[nodiscard]] bool
+RocksDBInvertedListsFilteringIterator<Strategy>::is_available() const {
   return _filteredIdsIt != _filteredIds.end() ||
-         (_batchIt->Valid() &&
-          _batchIt->key().starts_with(_rocksdbKey.string()));
+         (_it->Valid() && _it->key().starts_with(_rocksdbKey.string()));
 }
 
-bool RocksDBInvertedListsFilteringIterator::searchFilteredIds() {
+template<VectorIndexStoredValuesStrategy Strategy>
+bool RocksDBInvertedListsFilteringIterator<Strategy>::searchFilteredIds() {
   // Get documents ids from the vector index
   std::vector<LocalDocumentId> ids;
   std::unordered_map<LocalDocumentId, std::vector<uint8_t>> idsToValue;
   idsToValue.reserve(kBatchSize);
   ids.reserve(kBatchSize);
 
-  for (size_t i{0}; i < kBatchSize && _batchIt->Valid() &&
-                    _batchIt->key().starts_with(_rocksdbKey.string());
-       ++i, _batchIt->Next()) {
-    auto const id =
-        LocalDocumentId(RocksDBKey::indexDocumentId(_batchIt->key()));
+  for (size_t i{0}; i < kBatchSize && _it->Valid() &&
+                    _it->key().starts_with(_rocksdbKey.string());
```

Comment:
Here we also have the expression `_it->Valid() && _it->key().starts_with(_rocksdbKey.string()` which would match `RocksDBInvertedListsIteratorBase::is_available()`.